programming_examples/channel_examples/hierarchical_alloc/hierarchical_alloc.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (C) 2024, Advanced Micro Devices, Inc.
+# SPDX-License-Identifier: MIT
+
+from air.ir import *
+from air.dialects.air import *
+from air.dialects.memref import AllocOp, DeallocOp, load, store
+from air.dialects.func import FuncOp
+from air.dialects.scf import for_, yield_
+
+range_ = for_
+
+IMAGE_WIDTH = 1
+IMAGE_HEIGHT = 4
+IMAGE_SIZE = [IMAGE_WIDTH, IMAGE_HEIGHT]
+
+
+@module_builder
+def build_module():
+    memrefTyInOut = MemRefType.get(IMAGE_SIZE, T.i32())
+
+    mem_space_l1 = IntegerAttr.get(T.i32(), MemorySpace.L1)
+    mem_space_l2 = IntegerAttr.get(T.i32(), MemorySpace.L2)
+    mem_space_l3 = IntegerAttr.get(T.i32(), MemorySpace.L3)
+
+    image_type_l1 = MemRefType.get(
+        shape=IMAGE_SIZE,
+        element_type=T.i32(),
+        memory_space=mem_space_l1,
+    )
+    image_type_l2 = MemRefType.get(
+        shape=IMAGE_SIZE,
+        element_type=T.i32(),
+        memory_space=mem_space_l2,
+    )
+    image_type_l3 = MemRefType.get(
+        shape=IMAGE_SIZE,
+        element_type=T.i32(),
+        memory_space=mem_space_l3,
+    )
+
+    ChannelOp("ChanInL2")
+    ChannelOp("ChanOutL2")
+    ChannelOp("ChanInL1")
+    ChannelOp("ChanOutL1")
+
+    # We will send an image worth of data in and out
+    @FuncOp.from_py_func(memrefTyInOut, memrefTyInOut)
+    def copy(arg0, arg1):
+
+        # The arguments are the input and output
+        @launch(operands=[arg0, arg1])
+        def launch_body(a, b):
+            image_in_l3 = AllocOp(image_type_l3, [], [])
+
+            # Do some data manipulation instead of a straight copy
+            # 0, 1, 2, 3 -> 1, 1, 2, 2
+            c0 = arith.constant(T.index(), 0)
+            c1 = arith.constant(T.index(), 1)
+            c2 = arith.constant(T.index(), 2)
+            c3 = arith.constant(T.index(), 3)
+            val1 = load(a, [c0, c1])
+            store(val1, image_in_l3, [c0, c0])
+            store(val1, image_in_l3, [c0, c1])
+            val2 = load(a, [c0, c2])
+            store(val2, image_in_l3, [c0, c2])
+            store(val2, image_in_l3, [c0, c3])
+
+            ChannelPut("ChanInL2", image_in_l3)
+            DeallocOp(image_in_l3)
+
+            ChannelGet("ChanOutL2", b)
+
+            @segment(name="seg")
+            def segment_body():
+                image_in_l2a = AllocOp(image_type_l2, [], [])
+                image_in_l2b = AllocOp(image_type_l2, [], [])
+
+                ChannelGet("ChanInL2", image_in_l2a)
+
+                # Do some data manipulation instead of a straight copy
+                # 1, 1, 2, 2 -> 2, 1, 2, 2
+                c0 = arith.constant(T.index(), 0)
+                c1 = arith.constant(T.index(), 1)
+                c2 = arith.constant(T.index(), 2)
+                c3 = arith.constant(T.index(), 3)
+                val1 = load(image_in_l2a, [c0, c0])
```

Comment:
Similarly, program inside segment and outside of herd is going to be mapped to memtile DMA BDs. Memtile doesn't really have a core, so who will be doing the load or store?